datadog_checks_base/datadog_checks/base/utils/db/statement_metrics.py

```diff
@@ -0,0 +1,119 @@
+# (C) Datadog, Inc. 2020-present
+# All rights reserved
+# Licensed under a 3-clause BSD style license (see LICENSE)
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+class StatementMetrics:
+    """
+    This class supports normalized statement-level metrics, which are collected from the database's
+    statistics tables, ex:
+
+        - Postgres: pg_stat_statements
+        - MySQL: performance_schema.events_statements_summary_by_digest
+        - Oracle: V$SQLAREA
+        - SQL Server: sys.dm_exec_query_stats
+        - DB2: mon_db_summary
+
+    These tables are monotonically increasing.
+    """
+
+    def __init__(self):
+        self.previous_statements = dict()
+
+    def compute_derivative_rows(self, rows, metrics, key):
+        """
+        Compute the first derivative of column-based metrics for a given set of rows. This also resets the
+        statement cache so should only be called once per check run.
+
+        - **rows** (_List[dict]_) - rows from current check run
+        - **previous** (_List[dict]_) - rows from the previous check run
+        - **metrics** (_List[str]_) - the metrics to compute for each row
+        - **key** (_callable_) - function for an ID which uniquely identifies a row across runs
+        """
+        result = []
+        new_cache = {}
+        metrics = set(metrics)
+        dropped_metrics = set()
+        for row in rows:
+            row_key = key(row)
+            if row_key in new_cache:
+                logger.debug(
+                    'Collision in cached query metrics. Dropping existing row, row_key=%s new=%s dropped=%s',
+                    row_key,
+                    row,
+                    new_cache[row_key],
+                )
+            new_cache[row_key] = row
+            prev = self.previous_statements.get(row_key)
+            if prev is None:
+                continue
+            metric_columns = metrics & set(row.keys())
+            dropped_metrics.update(metrics - metric_columns)
+            if any([row[k] - prev[k] < 0 for k in metric_columns]):
+                # The table was truncated or stats reset; begin tracking again from this point
+                continue
```

Comment:
You're right, good catch! 🙏

Reply:
Actually a break would introduce its own bugs because all previous entries would need to be discarded. I updated and added tests to capture both cases.